### PR TITLE
feature/pfqueue-distributed

### LINKED ISSUE #2858: pfqueue: Provide ability to distribute some tasks
Right now when enqueuing a pfqueue task, the only option is to enqueue it on localhost.

It would be nice to be able to round-robin across the members of a cluster with a special option in the client or by using a special client (encapsulate the enqueue in a webservices call)

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Pfqueue distributed tasks

Impacts
=======
pfqueue and pfdchplistener

Issue
=====
fixes #2858

Delete branch after merge
=========================
NO

NEWS file entries
=================

Enhancements
------------
* Distribute pfdhcplistener task among the cluster memebers

### CODE REVIEW

File: conf/switches.conf.defaults [hunk 1]
```diff
@@ -29,7 +29,7 @@ VlanMap = Y
 RoleMap = N
 ExternalPortalEnforcement = N
 
-mode = testing
+mode = production
```
Comment (julsemaan) [inline conf/switches.conf.defaults]: Does this really belong in this PR ?

From what I know this is already in devel

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 2]
```diff
@@ -146,12 +146,12 @@ The following provides a list of the minimum server hardware recommendations:
 * Intel or AMD CPU 3 GHz
 * 8 GB of RAM
 * 100 GB of disk space (RAID-1 recommended)
-* 1 Network card (2 recommended)
+* 1 network card (2 recommended)
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: same comment as above for this change

**Comment by julsemaan** [comment]
This would need more verbose logging of servers being declared dead/alive and some debug of where the packets are being sent so its easier to troubleshoot

File: conf/switches.conf.defaults (hunk 1 above)
Comment (jrouzierinverse) [reply conf/switches.conf.defaults]: Fixed

**Comment by jrouzierinverse** [comment]
Come to think about we should also have a very fast timeout.
So we can fail fast

File: lib/pf/api/jsonrpcclient.pm [hunk 3]
```diff
@@ -83,6 +83,22 @@ has port => (is => 'rw', default => sub {$Config{'webservices'}{'port'}} );
...
+
+Curl connection timeout in milli seconds
+
+=cut
+
+has connect_timeout_ms => (is => 'rw', default => sub {0}) ;
+
+=head2 timeout_ms
+
+Curl transfer timeout in milli seconds
```
Comment (julsemaan) [inline lib/pf/api/jsonrpcclient.pm:105]: Not 100% sure but I believe there is no space between milli and seconds

File: lib/pf/api/jsonrpcclient.pm [hunk 4]
```diff
@@ -177,6 +197,8 @@ sub curl {
     $curl->setopt(CURLOPT_NOSIGNAL, 1);
     $curl->setopt(CURLOPT_URL, $url);
     $curl->setopt(CURLOPT_HTTPHEADER, ['Content-Type: application/json-rpc',"Request: $function"]);
+    $curl->setopt(CURLOPT_CONNECTTIMEOUT_MS, $self->connect_timeout_ms // 0);
```
Comment (julsemaan) [inline lib/pf/api/jsonrpcclient.pm:215]: It already defaults to 0.

Either 0 should be a constant used here and in the default value above or this should be removed so its not duplicate

File: lib/pf/api/jsonrpcclient.pm [hunk 5]
```diff
@@ -177,6 +197,8 @@ sub curl {
...
+    $curl->setopt(CURLOPT_TIMEOUT_MS, $self->timeout_ms // 0);
```
Comment (julsemaan) [inline lib/pf/api/jsonrpcclient.pm:216]: Same as my comment above

File: lib/pf/api/queue_cluster.pm [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+=head2 _build_local_client
+
+Build the local client
+
+=cut
+
+sub _build_local_client {
+    my ($self) = @_;
+    require pf::pfqueue::producer::redis;
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: Any reason why you aren't "using" the module ? That would be more effective I believe

File: lib/pf/api/queue_cluster.pm [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        $self->cluster_notify($method, @args);
+    } else {
+        $self->local_notify($method, @args);
+    }
+    return;
+}
+
+=head2 local_notify
+
+local_notify
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: A bit more descriptive information would help

File: lib/pf/api/queue_cluster.pm [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+=cut
+
+sub local_notify {
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: perldoc

File: lib/pf/api/queue_cluster.pm [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    my ($self, $method, @args) = @_;
+    get_logger->debug("sending $method locally");
+    $self->local_client->submit($self->queue, api => [$method, @args]);
+    return;
+}
+
+sub cluster_notify {
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: perldoc

File: lib/pf/api/queue_cluster.pm [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+Checks if server is down
+
+=cut
+
+sub is_server_down {
+    my ($server) = @_;
+    return $CHI_CACHE->get($server->{management_ip});
+}
+
+sub notify_delayed {
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: perldoc

File: lib/pf/api/queue_cluster.pm [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+=cut
+
+sub server_notify_delayed {
+    my ($self, $server, $delay, $method, @args) = @_;
+    return $self->do_jsonrpc_notify($server, queue_submit_delayed => $self->queue, $delay, api => [$method, @args]);
+}
+
+=head2 do_jsonrpc_notify
+
+do_jsonrpc_notify
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm:216]: Expand on this

File: lib/pf/api/queue_cluster.pm [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+            get_logger->info("sent $method to $server->{host}");
+            last;
+        }
+    }
+    return; 
+}
+
+=head2 server_notify_delayed
+
+server_notify_delayed
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm:205]: Expand

File: lib/pf/api/queue_cluster.pm [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        get_logger->error("Failed to send $method to $server->{host} marking as dead for $dead_for seconds");
+        $self->mark_server_as_down($server);
+    }
+
+    return $results;
+}
+
+=head2 jsonrpcclient
+
+jsonrpcclient
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm:235]: More meat please

image

File: lib/pf/api/queue_cluster.pm [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+
+=cut
+
+sub jsonrpcclient {
+    my ($self, $server) = @_;
+    require pf::api::jsonrpcclient;
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: Use vs require, that is the question

File: lib/pf/api/queue_cluster.pm [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+    return pf::api::jsonrpcclient->new(
+        host => $server->{management_ip},
+        proto => 'https',
+        %{$self->jsonrpc_args // {}},
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: Already defaulting to '{}'

File: lib/pf/api/queue_cluster.pm [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+
+sub local_notify_delayed {
+    my ($self, $delay, $method, @args) = @_;
+    $self->client->submit_delayed($self->queue, 'api', $delay, [$method, @args]);
+    return;
+}
+
+=head2 servers
+
+servers
```
Comment (julsemaan) [inline lib/pf/api/queue_cluster.pm]: needs = description++

zammit = pain + dinde + moutarde

File: lib/pf/util.pm [hunk 17]
```diff
@@ -940,6 +941,24 @@ sub trim_path {
    return ((@parts == 0) ? '' : catdir(@parts));
 }
 
+
+=item expand_csv
+
```
Comment (julsemaan) [inline lib/pf/util.pm:952]: Description needed

**Comment by julsemaan** [comment]
As a side note, I've tested this in a high throughput environment and its holding up so the code works

**Comment by jrouzierinverse** [comment]
Is this for 7.4 or 8.0?

**Comment by julsemaan** [comment]
That would be 8.0 I believe, the merge window for features of 7.4 is long gone and I don't think we'll be able to test it correctly and document the current limitations (like parking)

**Comment by extrafu** [comment]
@julsemaan Since it was tested in production environments already, should we just merge it to get it in v8 with 6 more weeks of testing?

**Comment by julsemaan** [comment]
@extrafu 

Sure, we just need @jrouzierinverse to adjust the PR based on the comments that were made and we're good for merge

**Comment by extrafu** [comment]
@jrouzierinverse Do the last changes and inform us when done. We need to merge this.

File: lib/pf/api/jsonrpcclient.pm (hunk 4 above)
Comment (jrouzierinverse) [reply lib/pf/api/jsonrpcclient.pm:215]: That is there to if connect_timeout_ms was set to undef.

File: lib/pf/api/jsonrpcclient.pm (hunk 5 above)
Comment (jrouzierinverse) [reply lib/pf/api/jsonrpcclient.pm:216]: That is there to if timeout_ms was set to undef.

File: lib/pf/api/queue_cluster.pm (hunk 15 above)
Comment (jrouzierinverse) [reply lib/pf/api/queue_cluster.pm]: That is there to if jsonrpc_args was set to undef.

**Comment by jrouzierinverse** [comment]
Done

**Comment by extrafu** [comment]
@julsemaan If all is good for you, please merge.

**Pull request merged**